### ISSUE TITLE
Use of static inline functions in headers may violate ODR and lead to UB

### ISSUE THREAD

**Issue by noloader** [weidai11/cryptopp#127]
This caught my eye from the CFE-Users mailing list: [static inline functions in headers and -Wunused-function](http://lists.llvm.org/pipermail/cfe-users/2016-January/000854.html) (Thread view available at [January 2016 Archives by thread](http://lists.llvm.org/pipermail/cfe-users/2016-January/thread.html)). It caught my eye because of ***`StringNarrow`***.

If I am parsing the thread correctly, then ***`static inline`*** should not be used on a function in a header *when* the header includes the definition.

According to David Blaikie, who answered the question and is likely a LLVM dev:

> They [static inline functions in the header] should probably just be "inline" without the static.
>
>static functions are distinct functions per translation unit - so having static functions in headers has a number of problems:
>
> * The function definitions won't be deduplicated by the linker across object files -> code bloat (larger binaries) & address inequality
> * This can lead to technical (& undiagnosed) Undefined Behavior due to ODR violations
>
> Consider the following header:
> 
>      static inline void f() { }
>      struct foo {
>        void g() { f(); }
>      };
>
> Include that header in two .cpp files, compile and link the two objects together and the program now violates the ODR..

-----

A quick audit reveals the following.

```
$ grep static * | grep inline | grep -v "\.cpp"
idea.h:		static inline void LookupMUL(word &a, word b);
iterhash.h:	inline static void CorrectEndianess(HashWordType *out, const HashWordType *in, size_t byteCount)
misc.h:static inline std::string StringNarrow(const wchar_t *str, bool throwOnError = true)
misc.h:	static inline GetBlock<T, B, GA> Get(const void *block) {return GetBlock<T, B, GA>(block);}
misc.h:	static inline T RightShift(T value, unsigned int bits)
misc.h:	static inline T LeftShift(T value, unsigned int bits)
misc.h:	static inline T RightShift(T value, unsigned int bits)
misc.h:	static inline T LeftShift(T value, unsigned int bits)
```

-----

Also see the following threads on the User Group:

* [cryptopp/misc.h:548:20: warning: unused function 'StringNarrow' [-Wunused-function]](http://groups.google.com/forum/#!topic/cryptopp-users/ySAaAdTPGwY)
* [Use of static inline functions in headers may violate ODR and lead to UB (Issue 127)](http://groups.google.com/forum/#!topic/cryptopp-users/0dW8ghx2vdQ)

**Comment by noloader** [comment]
Below are the results of the review.

-----

***(1)***

>     idea.h: static inline void LookupMUL(word &a, word b);

OK.

Member function, part of `class IDEA`. Implementation is in source (*.CPP) file.

-----

***(2)***

>     iterhash.h: inline static void CorrectEndianess(HashWordType *out, const HashWordType *in, size_t byteCount)

Member function, part of `class IteratedHash`. Implementation is in header (*.H) file.

-----

***(3)***

>     misc.h:static inline std::string StringNarrow(const wchar_t *str, bool throwOnError = true)

Non-member function. Implementation is in header (*.H) file. It looks like an ODR violation.

-----

***(4)***

**Comment by noloader** [comment]
Here's the patch to remediate it the issue with `StringNarrow`:

```
$ cat git.diff
diff --git a/misc.cpp b/misc.cpp
index b5a2ea4..8d2ccb8 100644
--- a/misc.cpp
+++ b/misc.cpp
@@ -133,6 +133,64 @@ bool VerifyBufsEqual(const byte *buf, const byte *mask, size_t count)
 	return acc8 == 0;
 }
 
+#ifndef CRYPTOPP_MAINTAIN_BACKWARDS_COMPATIBILITY_562
+std::string StringNarrow(const wchar_t *str, bool throwOnError)
+{
+	assert(str);
+	std::string result;
+
+	// Safer functions on Windows for C&A, https://github.com/weidai11/cryptopp/issues/55
+#if (CRYPTOPP_MSC_VERSION >= 1400)
+	size_t len=0, size = 0;
+	errno_t err = 0;
+
+	//const wchar_t* ptr = str;
+	//while (*ptr++) len++;
+	len = wcslen(str)+1;
+
+	err = wcstombs_s(&size, NULL, 0, str, len*sizeof(wchar_t));
+	assert(err == 0);
+	if (err != 0) {goto CONVERSION_ERROR;}
+
+	result.resize(size);
+	err = wcstombs_s(&size, &result[0], size, str, len*sizeof(wchar_t));
+	assert(err == 0);
+
+	if (err != 0)
+	{
+CONVERSION_ERROR:
+		if (throwOnError)
+			throw InvalidArgument("StringNarrow: wcstombs_s() call failed with error " + IntToString(err));
+		else
+			return std::string();
+	}
+
+	// The safe routine's size includes the NULL.
+	if (!result.empty() && result[size - 1] == '\0')
+		result.erase(size - 1);
+#else
+	size_t size = wcstombs(NULL, str, 0);
+	assert(size != (size_t)-1);
+	if (size == (size_t)-1) {goto CONVERSION_ERROR;}
+
+	result.resize(size);
+	size = wcstombs(&result[0], str, size);
+	assert(size != (size_t)-1);
+
+	if (size == (size_t)-1)
+	{
+CONVERSION_ERROR:
+		if (throwOnError)
+			throw InvalidArgument("StringNarrow: wcstombs() call failed");
+		else
+			return std::string();
+	}
+#endif
+
+	return result;
+}
+#endif // StringNarrow and CRYPTOPP_MAINTAIN_BACKWARDS_COMPATIBILITY_562
+
 #if !(defined(_MSC_VER) && (_MSC_VER < 1300))
 using std::new_handler;
 using std::set_new_handler;

diff --git a/misc.h b/misc.h
index 341cf28..26c620c 100644
--- a/misc.h
+++ b/misc.h
@@ -1072,10 +1072,9 @@ inline void SecureWipeArray(T *buf, size_t n)
 //!   (0xE9 0xAA 0xA8), then you should ensure the locales are available. If the locales are not available,
 //!   then a 0x21 error is returned which eventually results in an InvalidArgument exception
 #ifndef CRYPTOPP_MAINTAIN_BACKWARDS_COMPATIBILITY_562
-static inline std::string StringNarrow(const wchar_t *str, bool throwOnError = true)
+inline std::string StringNarrow(const wchar_t *str, bool throwOnError = true);
 #else
 static std::string StringNarrow(const wchar_t *str, bool throwOnError = true)
-#endif
 {
 	assert(str);
 	std::string result;
@@ -1093,7 +1092,7 @@ static std::string StringNarrow(const wchar_t *str, bool throwOnError = true)
 	assert(err == 0);
 	if (err != 0) {goto CONVERSION_ERROR;}
 
-	result.resize(size);	
+	result.resize(size);
 	err = wcstombs_s(&size, &result[0], size, str, len*sizeof(wchar_t));
 	assert(err == 0);
 
@@ -1113,7 +1112,7 @@ CONVERSION_ERROR:
 	size_t size = wcstombs(NULL, str, 0);
 	assert(size != (size_t)-1);
 	if (size == (size_t)-1) {goto CONVERSION_ERROR;}
-	
+
 	result.resize(size);
 	size = wcstombs(&result[0], str, size);
 	assert(size != (size_t)-1);
@@ -1130,6 +1129,7 @@ CONVERSION_ERROR:
 
 	return result;
 }
+#endif // StringNarrow and CRYPTOPP_MAINTAIN_BACKWARDS_COMPATIBILITY_562
```

**Comment by noloader** [comment]
Fixed in [Commit 3a3fae8b8b2139db](http://github.com/weidai11/cryptopp/commit/3a3fae8b8b2139dbcd6ac359bb76f553ba8bf01f).
